crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -0,0 +1,147 @@
+use {
+    super::{NativePriceEstimateResult, NativePriceEstimating},
+    crate::price_estimation::PriceEstimationError,
+    anyhow::{anyhow, Result},
+    futures::{future::BoxFuture, FutureExt},
+    primitive_types::H160,
+    reqwest::{Client, StatusCode},
+    serde::Deserialize,
+    std::collections::HashMap,
+    url::Url,
+};
+
+#[derive(Debug, Deserialize)]
+struct Response(HashMap<H160, Price>);
+
+#[derive(Debug, Deserialize)]
+struct Price {
+    eth: f64,
+}
+
+type Token = H160;
+
+pub struct CoinGecko {
+    client: Client,
+    base_url: Url,
+    api_key: Option<String>,
+    chain: String,
+}
+
+impl CoinGecko {
+    /// Authorization header for CoinGecko
+    const AUTHORIZATION: &'static str = "x-cg-pro-api-key";
+
+    pub fn new(
+        client: Client,
+        base_url: Url,
+        api_key: Option<String>,
+        chain_id: u64,
+    ) -> Result<Self> {
+        let chain = match chain_id {
+            1 => "ethereum".to_string(),
+            100 => "xdai".to_string(),
+            42161 => "arbitrum-one".to_string(),
+            n => anyhow::bail!("unsupported network {n}"),
+        };
+        Ok(Self {
+            client,
+            base_url,
+            api_key,
+            chain,
+        })
+    }
+}
+
+impl NativePriceEstimating for CoinGecko {
+    fn estimate_native_price(&self, token: Token) -> BoxFuture<'_, NativePriceEstimateResult> {
+        async move {
+            let url = format!(
+                "{}/{}?contract_addresses={token:#x}&vs_currencies=eth",
+                self.base_url, self.chain
+            );
+            let mut builder = self.client.get(&url);
+            if let Some(ref api_key) = self.api_key {
+                builder = builder.header(Self::AUTHORIZATION, api_key)
+            }
+            observe::coingecko_request(&url);
+            let response = builder.send().await;
+            observe::coingecko_response(&url, response.as_ref());
```

Comment:
nit: Not sure how this log looks but it likely doesn't contain body (since I believe this needs an additional await). I believe it's useful to print the response body in tracing to see (e.g. if the format changes) what the API returned to us concretely.